docs/design/readme.md

```diff
@@ -0,0 +1,1704 @@
+        {
+            return unexpected(res.error());
+        }
+        return publish(deadline, buf);
+    }
+};
+```
+
+Contrary to the [original execution flow design proposal](https://forum.opencyphal.org/t/design-review-execution-model-for-libcyphal/1918), this design intentionally chooses to not queue the serialized message(s) inside the publisher instance until the next `run`, as that would carry the following disadvantages:
+
+- The serialization buffer will have to be kept off the stack as it has to outlive the stack frame of the `publish` method. This can be achieved either by allocating the buffer on the stack and then moving it to the `PublisherBase` instance for keeping until the next `run`, or by continuously storing a pre-allocated buffer as a field inside `Publisher<Message>`. The first approach requires dynamic memory which is undesirable, the second approach requires pre-allocating a potentially large buffer, which is wasteful.
+
+- Message publications will be delayed until the next `run` cycle.
+
+One potential downside of the current proposal is that the publication call stack may be quite deep, going all the way from the application into a lizard. This is especially important when redundant transports are used.
+
+The proposal intentionally excludes statistical counters; such auxiliary features are to be retrofitted at a later stage.
+
+### Subscriber
+
+```c++
+class SubscriberBase
+{
+    friend class SubscriberImpl;  //< Friendship is needed to allow invocation of accept().
+
+public:
+    struct Metadata final : public transport::TransferMetadata  // TODO Find a better place for this; there's some reuse potential
+    {
+        std::optional<std::uint16_t> publisher_node_id;
+    };
+
+    [[nodiscard]] transport::IMessageRxSession& getSession() noexcept;
+    [[nodiscard]] const transport::IMessageRxSession& getSession() const noexcept;
+
+    /// The transfer-ID timeout is shared between all subscriber objects on this port.
+    /// The default is 2 seconds.
+    [[nodiscard]] Duration getTransferIDTimeout() const noexcept { return impl_->getTransferIDTimeout(); }
+    void setTransferIDTimeout(const Duration dur) { impl->setTransferIDTimeout(dur); }
+
+    // TODO: Add statistical counters: number of transfers received, deserialization errors, queue overruns.
+
+    // The type is movable and non-copyable. Moving the instance causes the reference stored in SubscriberImpl to be updated.
+    // Alternatively, the type can be made non-movable to reduce complexity.
+    SubscriberBase(const SubscriberBase&)            = delete;
+    SubscriberBase(SubscriberBase&&)                 = default;
+    SubscriberBase& operator=(const SubscriberBase&) = delete;
+    SubscriberBase& operator=(SubscriberBase&&)      = default;
+    virtual ~SubscriberBase() noexcept               = default;
+
+protected:
+    explicit SubscriberBase(const SharedPtr<PublisherImpl>& impl);
```

Comment:
Should this be  `SharedPtr<SubscriberImpl>` instead of  `SharedPtr<PublisherImpl>`?